src/main/java/com/powsybl/caseserver/datasource/S3CaseDataSourceService.java

```diff
@@ -51,34 +53,30 @@ public Boolean datasourceExists(UUID caseUuid, String fileName) {
 
     @Override
     public byte[] getInputStream(UUID caseUuid, String fileName) {
-        final var caseFileKey = Objects.nonNull(s3CaseService.getCompressionFormat(caseUuid)) && s3CaseService.getCompressionFormat(caseUuid).equals("zip")
-                ? uuidToPrefixKey(caseUuid) + fileName
-                : uuidToPrefixKey(caseUuid) + s3CaseService.getCaseName(caseUuid);
+        String caseName = s3CaseService.getCaseName(caseUuid);
+        String caseFileKey;
+        // For archived cases (.zip, .tar, ...), individual files are gzipped in S3 server.
+        // Here the requested file is decompressed and simply returned.
+        if (S3CaseService.isArchivedCaseFile(caseName)) {
+            caseFileKey = uuidToKeyWithFileName(caseUuid, fileName + GZIP_EXTENSION);
+            return s3CaseService.withS3DownloadedTempPath(caseUuid, caseFileKey,
+                    file -> S3CaseService.decompress(Files.readAllBytes(file)));
+        }
+        caseFileKey = uuidToKeyWithFileName(caseUuid, caseName);
         return withS3DownloadedDataSource(caseUuid, caseFileKey,
-            datasource -> IOUtils.toByteArray(datasource.newInputStream(Paths.get(fileName).getFileName().toString())));
-    }
-
-    private String uuidToPrefixKey(UUID uuid) {
-        return CASES_PREFIX + uuid.toString() + "/";
+                datasource -> IOUtils.toByteArray(datasource.newInputStream(Paths.get(fileName).getFileName().toString())));
     }
 
     @Override
     public byte[] getInputStream(UUID caseUuid, String suffix, String ext) {
-        return withS3DownloadedDataSource(caseUuid,
-            datasource -> IOUtils.toByteArray(datasource.newInputStream(suffix, ext)));
+        return getInputStream(caseUuid, DataSourceUtil.getFileName(getBaseName(caseUuid), suffix, ext));
     }
 
     @Override
     public Set<String> listName(UUID caseUuid, String regex) {
         return s3CaseService.listName(caseUuid, regex);
     }
 
-    public <R, T extends Throwable> R withS3DownloadedDataSource(UUID caseUuid, FailableFunction<DataSource, R, T> f) {
-        FailableFunction<Path, DataSource, T> pathToDataSource = DataSource::fromPath;
-        FailableFunction<Path, R, T> composedFunction = pathToDataSource.andThen(f);
-        return s3CaseService.withS3DownloadedTempPath(caseUuid, composedFunction);
-    }
-
     public <R, T extends Throwable> R withS3DownloadedDataSource(UUID caseUuid, String caseFileKey, FailableFunction<DataSource, R, T> f) {
```

Comment:
not sure if this one is needed anymore, we should never download the whole datasource now

Reply:
Yes both withS3DownloadedDataSource can be deleted